coalib/collecting/Collectors.py

```diff
@@ -54,11 +57,57 @@ def icollect(file_paths, ignored_globs=None, match_cache={}):
             match_cache[file_path] = list(iglob(file_path))
 
         for match in match_cache[file_path]:
-            if not ignored_globs or not fnmatch(match, ignored_globs):
+            if not match_function(match, ignored_globs):
                 yield match, file_path
 
 
-def collect_files(file_paths, log_printer=None, ignored_file_paths=None,
+def partition(iterable, predicate):
+    """
+    Partitions the iterable into two iterables based on the given predicate
+
+    :param predicate:   A function that takes an item of the iterable and
+                        returns a boolean
+    :return:            Two iterators pointing to the original iterable
+    """
+    a, b = itertools.tee(((predicate(item), item) for item in iterable))
+
+    return ([item for pred, item in a if pred],
+            [item for pred, item in b if not pred])
+
+
+def match_dir_or_file_pattern(path, file_patterns=(), directory_patterns=()):
+    """
+    Tries to match the given path with the directory (substring match) or file
+    (enforced full match) patterns.
+    Returns true if any of the given pattern match.
+
+    :param path:                Valid file path
+    :param file_patterns:       List of regex patterns that match a file
+    :param directory_patterns:  List of regex patterns that match a directory
+    """
+    return (
+        any(re.match(pattern.replace('\\', '\\\\'), path)
+            for pattern
+            in directory_patterns) or
+        any(re.fullmatch(pattern.replace('\\', '\\\\'), path)
+            for pattern
+            in file_patterns))
+
+
+def list_glob_results(values):
+    """
+    Expands the globs of all given values and concatenates the results.
+
+    :param values:  List of file-globs or files.
+    :return:        List of matched files.
+    """
+    return functools.reduce(
+        lambda seed, value: seed + list(iglob(value)),
+        values,
+        [])
+
+
+def collect_files(file_paths, log_printer=None, ignored_file_paths=(),
```

Comment:
`ignored_file_paths=()` changes the semantics of the function, which is why you had to change some of the existing tests.

It is generally a bad idea to rely on args default being a empty collection.  The default should be None, and inside the function it does the right thing for any False-ish value.

i.e., inside, add `if not foo: foo = (, )` , as this is an internal matter.

As a general rule, a function should never distinguish between different types of False-ish provided in data given in arguments unless it was part of a very clear contract.

Reply:
Can you explain why this is generally bad?
I understand that argument in case of empty list defaults, because the list is only allocated once and all changes values will carry over to the next function call.
However, this is not the case with tuples.

The advantage of `None` is that we are forced to check for Falsy-Values.
However, we also accept all other Falsy values, which can lead to weird behaviour, like accepting `0` and throwing when another number is provided.

In my opinion, using a tuple default value if a collection is expected makes the contract a lot clearer for people using the function.